sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/AuthClient.java

```diff
@@ -0,0 +1,182 @@
+            resource, tenantId, clientId, clientSecret });
+        LOGGER.info("Getting access token using client ID / client secret");
+        String result = null;
+
+        StringBuilder oauth2Url = new StringBuilder();
+        oauth2Url.append(OAUTH2_TOKEN_BASE_URL)
+                 .append(tenantId)
+                 .append(OAUTH2_TOKEN_POSTFIX);
+
+        StringBuilder requestBody = new StringBuilder();
+        requestBody.append(GRANT_TYPE_FRAGMENT)
+                   .append(CLIENT_ID_FRAGMENT).append(clientId)
+                   .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
+                   .append(RESOURCE_FRAGMENT).append(resource);
+
+        String body = post(oauth2Url.toString(), requestBody.toString(), "application/x-www-form-urlencoded");
+        if (body != null) {
+            JsonConverter converter = JsonConverterFactory.createJsonConverter();
+            OAuthToken token = (OAuthToken) converter.fromJson(body, OAuthToken.class);
+            result = token.getAccess_token();
+        }
+        LOGGER.log(FINER, "Access token: {0}", result);
+        return result;
+    }
+
+    /**
+     * Get the access token on Azure App Service.
+     *
+     * @param resource the resource.
+     * @return the authorization token.
+     */
+    private String getAccessTokenOnAppService(String resource) {
+        LOGGER.entering("AuthClient", "getAccessTokenOnAppService", resource);
+        LOGGER.info("Getting access token using managed identity based on MSI_SECRET");
+        String result = null;
+
+        StringBuilder url = new StringBuilder();
+        url.append(System.getenv("MSI_ENDPOINT"))
+           .append("?api-version=2017-09-01")
+           .append(RESOURCE_FRAGMENT).append(resource);
+
+        HashMap<String, String> headers = new HashMap<>();
+        headers.put("Metadata", "true");
+        headers.put("Secret", System.getenv("MSI_SECRET"));
+        String body = get(url.toString(), headers);
+
+        if (body != null) {
+            JsonConverter converter = JsonConverterFactory.createJsonConverter();
+            OAuthToken token = (OAuthToken) converter.fromJson(body, OAuthToken.class);
+            result = token.getAccess_token();
+        }
+        LOGGER.exiting("AuthClient", "getAccessTokenOnAppService", result);
+        return result;
+    }
+
+    /**
+     * Get the authorization token on everything else but Azure App Service.
+     *
+     * @param resource the resource.
+     * @return the authorization token.
+     */
+    private String getAccessTokenOnOthers(String resource) {
+        LOGGER.entering("AuthClient", "getAccessTokenOnOthers", resource);
+        LOGGER.info("Getting access token using managed identity");
+        String result = null;
+
+        StringBuilder url = new StringBuilder();
+        url.append(OAUTH2_MANAGED_IDENTITY_TOKEN_URL)
+           .append(RESOURCE_FRAGMENT).append(resource);
```

Comment:
I believe the resource string will need to be url encoded no?

Reply:
The KeyVaultClient code passes in a url encoded version so nothing further needs to be done